### PR TITLE
feat(lint): switch to ruff

### PR DESCRIPTION
**Pull request by onerandomusername**
## Summary

Migrate to ruff, and stop using flake8. This speeds up a full repository lint from 25+ seconds to half a second.

If we choose to use ruff's import sorting, then we can remove isort and speed up by about another second (of ~2.5)

## Checklist

- [ ] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by shiftinv — CHANGES_REQUESTED** [review]
~~how on earth does it run so quickly :o~~

File: pyproject.toml [hunk 1]
```diff
@@ -49,6 +49,73 @@ line_length = 100
...
+"test_bot/*.py" = ["B008", "T201", "PT"]
+"tests/*.py" = ["S101"]
+
+[tool.ruff.flake8-pytest-style]
+fixture-parentheses = false
+mark-parentheses = false
+
+[tool.ruff.flake8-tidy-imports]
+[tool.ruff.flake8-tidy-imports.banned-api]
+"subprocess".msg = "Consider possible security implications associated with the subprocess module." # replaces S404
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply pyproject.toml:118]: https://www.toml-lint.com/ uses an *old* parser, that is compatiable with v0.4.0 of the toml spec.

File: docs/extensions/attributetable.py [hunk 2]
```diff
@@ -1,20 +1,23 @@
...
-from sphinx.environment import BuildEnvironment
 from sphinx.locale import _
 from sphinx.util.docutils import SphinxDirective
-from sphinx.writers.html import HTMLTranslator
+
+if TYPE_CHECKING:
+    from _types import SphinxExtensionMeta
+    from sphinx.application import Sphinx
+    from sphinx.environment import BuildEnvironment
+    from sphinx.writers.html import HTMLTranslator
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply docs/extensions/attributetable.py:20]: this was the result of a partial attempt to use flake8-typechecking errors.

It flagged too many things that were not issues, but I made the changes here before removing that.

I'd like to keep the changes in the docs tho, if possible.

File: disnake/ext/commands/base_core.py [hunk 3]
```diff
@@ -627,8 +627,8 @@ async def can_run(self, inter: ApplicationCommandInteraction) -> bool:
             if not predicates:
                 # since we have no checks, then we just return True.
                 return True
-
-            return await async_all(predicate(inter) for predicate in predicates)  # type: ignore
+            else:
+                return await async_all(predicate(inter) for predicate in predicates)  # type: ignore
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/ext/commands/base_core.py:631]: i'll disable it for now because you're right. But we should consider using it in the future. IIRC there's a way that python can be hiding tracebacks with the `finally` block and returning within the `try:`

File: pyproject.toml [hunk 4]
```diff
@@ -49,6 +49,73 @@ line_length = 100
...
 
 [tool.taskipy.tasks]
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply pyproject.toml:118]: https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: examples/basic_voice.py [hunk 5]
```diff
@@ -73,7 +73,8 @@ async def join(self, ctx, *, channel: disnake.VoiceChannel):
         """Joins a voice channel"""
 
         if ctx.voice_client is not None:
-            return await ctx.voice_client.move_to(channel)
+            await ctx.voice_client.move_to(channel)
+            return
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply examples/basic_voice.py]: https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: pyproject.toml [hunk 6]
```diff
@@ -49,6 +49,73 @@ line_length = 100
...
+    "PGH", # pygrep-hooks
+    "TRY002", "TRY004","TRY300", # tryceratops
+]
+ignore = [
+    # star imports
+    "F401","F403",
+
+    # unknown if this is actually an issue
+    "RUF006",
+
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply pyproject.toml:85]: it is now 
https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: disnake/ext/commands/params.py [hunk 7]
```diff
@@ -629,8 +629,8 @@ async def convert_argument(self, inter: ApplicationCommandInteraction, argument:
             argument = self.converter(inter, argument)
             if inspect.isawaitable(argument):
                 return await argument
-
-            return argument
+            else:
+                return argument
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/ext/commands/params.py]: https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: disnake/ext/commands/core.py [hunk 8]
```diff
@@ -1149,8 +1149,8 @@ async def can_run(self, ctx: Context) -> bool:
             if not predicates:
                 # since we have no checks, then we just return True.
                 return True
-
-            return await disnake.utils.async_all(predicate(ctx) for predicate in predicates)  # type: ignore
+            else:
+                return await disnake.utils.async_all(predicate(ctx) for predicate in predicates)  # type: ignore
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/ext/commands/core.py]: https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: disnake/ext/commands/base_core.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (onerandomusername) [reply disnake/ext/commands/base_core.py]: https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -55,12 +55,8 @@ repos:
...
     hooks:
-      - id: flake8
-        name: flake8
-        description: "Lint all of our code"
-        entry: flake8
-        language: system
-        types: [python]
-        require_serial: true
+      - id: ruff
+        args: [--fix, --exit-non-zero-on-fix]
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply .pre-commit-config.yaml]: https://github.com/DisnakeDev/disnake/pull/935/commits/9d129c0c55bf70a38e9027238a59f303af968eeb

File: docs/extensions/attributetable.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shiftinv) [reply docs/extensions/attributetable.py:20]: Alright, fair enough. Would be great if `flake8-typechecking` worked, but there are just so many edge cases :/

File: pyproject.toml [hunk 10]
```diff
@@ -49,11 +49,75 @@ line_length = 100
...
+"examples/views/tic_tac_toe.py" = ["E741", "RET505"]
+"test_bot/*.py" = ["B008", "T201", "PT"]
+"tests/*.py" = ["S101"]
+
+[tool.ruff.flake8-pytest-style]
+fixture-parentheses = false
+mark-parentheses = false
+
+[tool.ruff.flake8-tidy-imports.banned-api]
+"subprocess".msg = "Consider possible security implications associated with the subprocess module." # replaces S404
```
Comment (shiftinv) [inline pyproject.toml:116]:
```suggestion
[tool.ruff.flake8-tidy-imports.banned-api.subprocess]
msg = "Consider possible security implications associated with the subprocess module." # replaces S404
```
This should make vscode's toml parser shut up.

Comment (onerandomusername) [reply pyproject.toml:116]: You should use a different plugin. Mine does not complain about this.

File: disnake/ext/commands/ctx_menus_core.py [hunk 11]
```diff
@@ -125,7 +125,7 @@ async def __call__(
         self, interaction: ApplicationCommandInteraction, target: Any = None, *args, **kwargs
     ) -> None:
         # the target may just not be passed in
-        args = (target or interaction.target,) + args
+        args = (target or interaction.target, *args)
```
Comment (shiftinv) [inline disnake/ext/commands/ctx_menus_core.py:128]: We should reconsider `RUF005`: https://github.com/charliermarsh/ruff/issues/2142#issuecomment-1402722345
While the changes themselves are generally fine, the speed increase is questionable and readability suffers sometimes.

**Comment by onerandomusername** [comment]
Keep in mind if #836 is merged first, https://github.com/DisnakeDev/disnake/pull/836#discussion_r1113136556 can be resolved here

File: tests/test_embeds.py [hunk 12]
```diff
@@ -11,7 +11,7 @@
 _BASE = {"type": "rich"}
 
 
-@pytest.fixture()
+@pytest.fixture
```
Comment (onerandomusername) [inline tests/test_embeds.py:14]: BTW, this is configurable, and we can go back to () if its desired

File: disnake/ext/commands/ctx_menus_core.py (hunk 11 above)
Comment (onerandomusername) [reply disnake/ext/commands/ctx_menus_core.py]: https://github.com/DisnakeDev/disnake/pull/935/commits/9885350a598bc314c3251c886ad797ea5c116815

File: tests/test_embeds.py (hunk 12 above)
Comment (shiftinv) [reply tests/test_embeds.py:14]: Seems fine, as long as it's consistent with other decorators.

File: pyproject.toml [hunk 13]
```diff
@@ -49,15 +49,94 @@ line_length = 100
 combine_as_imports = true
 filter_files = true
 
+[tool.ruff]
+line-length = 100
+target-version = "py38"
+src = ["disnake"]
```
Comment (shiftinv) [inline pyproject.toml:55]:
```suggestion
```
Probably doesn't matter, however `src` defaults to `["."]`, which is what we want.
https://beta.ruff.rs/docs/settings/#src

**Review by shiftinv — APPROVED** [review]
lgtm!

File: pyproject.toml (hunk 13 above)
Comment (onerandomusername) [reply pyproject.toml]: https://github.com/DisnakeDev/disnake/pull/935/commits/2a9430fb19dd528de7d1cca5667b561a4f9ad195
